config/src/network_id.rs

```diff
@@ -0,0 +1,87 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+use serde::{Deserialize, Serialize};
+
+#[derive(Clone, Debug, Deserialize, Eq, Hash, PartialEq, Serialize)]
+pub enum NetworkType {
+    Validator = 0,
+    Public,
+    Private,
+}
+
+/// A representation of the network being used in communication.
+/// There should only be one of each NetworkId used for a single node, and handshakes should verify
+/// that the NetworkId being used is the same during a handshake, to effectively ensure communication
+/// is restricted to a network.  Network should be checked that it is not the `DEFAULT_NETWORK`
+#[derive(Clone, Debug, Deserialize, Eq, Hash, PartialEq, Serialize)]
+pub struct NetworkId {
+    network_name: String,
+    network_type: NetworkType,
+}
+
+const INVALID: &str = "Invalid";
+
+impl Default for NetworkId {
+    fn default() -> NetworkId {
+        NetworkId::private_network(INVALID.to_string())
+    }
+}
+
+impl NetworkId {
+    /// Tell us whether the `NetworkId` was configured correctly.
+    /// Private networks must have a name, otherwise we can't differentiate.
+    pub fn is_valid(&self) -> bool {
+        match self.network_type {
+            NetworkType::Validator => true,
+            NetworkType::Public => true,
+            NetworkType::Private => self.network_name == INVALID,
+        }
+    }
+
+    pub fn private_network(name: String) -> NetworkId {
+        NetworkId {
+            network_type: NetworkType::Private,
+            network_name: name,
+        }
+    }
+
+    /// Public network to talk between Full nodes and Validators
+    pub fn public_network() -> NetworkId {
+        NetworkId {
+            network_type: NetworkType::Public,
+            network_name: "".to_string(),
+        }
+    }
+
+    /// Global Validator network
+    pub fn validator_network() -> NetworkId {
+        NetworkId {
+            network_type: NetworkType::Validator,
+            network_name: "".to_string(),
+        }
+    }
```

Comment:
your current approach requires this extra code.

Reply:
I can cut this out, the reason it required extra code was I found myself constantly switching between enums and structs trying to figure out which one was actually allowed in TOML.  By abstracting it away it let me change it easier.

I really wanted to use the enum approach, but TOML caused a whole bunch of serialization issues.